### PR TITLE
ADD: GasLib parsing functions and implementation of pressure reducing valves

### PR DESCRIPTION
**Pull request by tasseff**
This is a first shot at implementing a GasLib parser. It also includes the implementation of pressure reducing valves, which appear as special resistors in the new `GasLib-4197` case. The parser enables the solution of [all GasLib instances on the official website](http://gaslib.zib.de) when using the `gf` problem specification with the `MISOCPGasModel` formulation. I haven't yet tested with other problem specifications or formulations. A review would be appreciated!

### CODE REVIEW

File: src/core/constraint.jl [hunk 1]
```diff
@@ -32,6 +32,27 @@ function constraint_resistor_mass_flow(gm::AbstractGasModel, n::Int, k, f_min, f
...
 
+#################################################################################################
+# Constraints associated with prvs
+#################################################################################################
+
+"Constraint: Constraints which define pressure drop across a prv"
+function constraint_prv_pressure(gm::AbstractGasModel, n::Int, k, i, j, pd)
+    pi = var(gm, n, :psqr, i)
+    pj = var(gm, n, :psqr, j)
+    _add_constraint!(gm, n, :pressure_drop, k, JuMP.@constraint(gm.model, pd == pi - pj))
```
Comment (rb004f) [inline src/core/constraint.jl:43]: do prvs have a constant pressure drop?

File: src/core/constraint_template.jl [hunk 2]
```diff
@@ -61,6 +61,39 @@ function constraint_resistor_weymouth(gm::AbstractGasModel, k; n::Int=gm.cnw)
...
+
+"Template: Constraint on mass flow across a prv"
+function constraint_prv_mass_flow(gm::AbstractGasModel, k; n::Int=gm.cnw)
+    prv              = ref(gm, n, :prv, k)
+    f_min            = prv["flow_min"]
+    f_max            = prv["flow_max"]
+    is_bidirectional = get(prv, "is_bidirectional", 1)
+    flow_direction   = get(prv, "flow_direction", 0)
+
+    if is_bidirectional == 0 || flow_direction == 1
```
Comment (rb004f) [inline src/core/constraint_template.jl:76]: Take a look at how I am computing f_min and f_max now.  I am now doing this in _calc_component_max_flow... Basically, what I am storing in the ref is the tighted f_min and f_max bounds.  I found this logic was being done over and over again, so it made more sense to go ahead and do it in the calc function.

File: src/core/data.jl [hunk 3]
```diff
@@ -1125,6 +1137,20 @@ function _calc_resistor_resistance(resistor::Dict{String,Any})
 end
 
 
+"calculates the minimum flow on a prv"
+function _calc_prv_flow_min(ref::Dict{Symbol,Any}, prv)
+    mf = -ref[:max_mass_flow]
+    return max(mf, prv["flow_min"])
```
Comment (rb004f) [inline src/core/data.jl:1143]: This is where I am doing the tightening of the flow_min and flow_max.

File: test/data.jl [hunk 4]
```diff
@@ -1,4 +1,10 @@
 @testset "test data handling and parsing" begin
+    @testset "gaslib40 parsing from gaslib zip" begin
+        data = GasModels.parse_file("../test/data/gaslib/GasLib-40.zip")
+        @test length(data["delivery"]) == 29
+        @test length(data["receipt"]) == 3
```
Comment (rb004f) [inline test/data.jl:5]: somewhere can you also make a simple test of a model with the prv for the constraint?  I want to make sure that code is covered.

File: Project.toml [hunk 5]
```diff
@@ -10,9 +10,11 @@ Dierckx = "39dd38d3-220a-591b-8e3c-4c3a8c710a94"
 InfrastructureModels = "2030c09a-7f63-5d83-885d-db604e0e9cc0"
```
Comment (rb004f) [inline Project.toml:10]: can you add a note about adding support for prvs to the changelog?

File: Project.toml [hunk 6]
```diff
@@ -10,9 +10,11 @@ Dierckx = "39dd38d3-220a-591b-8e3c-4c3a8c710a94"
...
 JSON = "682c06a0-de6a-54ab-a142-c8b1cf79cde6"
 JuMP = "4076af6c-e467-56ae-b986-b466b2749572"
+ZipFile = "a5390f91-8eb1-5f08-bee0-b1d1ffed6cea"
 MathOptInterface = "b8f27783-ece8-5eb3-8dc8-9495eed66fee"
 Memento = "f28f55f0-a522-5efc-85c2-fe41dfb9b2d9"
 Printf = "de0858da-6303-5e67-8744-51eddeeeb8d7"
+XMLDict = "228000da-037f-5747-90a9-8195ccbf91a5"
```
Comment (rb004f) [inline Project.toml:18]: can you go through the docs and add documentation for the prvs in the appropriate places (mainly just the places where the components are discussed and the example problem definitions)

File: src/core/constraint.jl (hunk 1 above)
Comment (tasseff) [reply src/core/constraint.jl:43]: Yes, this is how these special resistors are defined in GasLib.

However, according to [this](https://en.wikipedia.org/wiki/Pressure_regulator), a pressure reducing valve only affects the pressure at the outlet node (`j`) and is insensitive to the pressure at the inlet node (`i`). Given that, maybe "pressure reducing valve" isn't the best nomenclature. Again, [GasLib documentation](https://www.mdpi.com/2306-5729/2/4/40/pdf) calls these components resistors (search for "pressureLoss"), which conflicts with how GasModels currently defines resistors. Maybe it would be better to rename "prv" to some special kind of resistor? I'm not quite sure.

Comment (rb004f) [reply src/core/constraint.jl:43]: Yes, that was the source of my confusion. When I looked at the constraint, it didn't seem like prv type constraint.  I would recommend calling it some variation of a resistor... does the gaslib documentation or papers refer to it in any special way?  The other thing we could do is "type" the resistor like we do with compressors

Comment (tasseff) [reply src/core/constraint.jl:43]: The GasLib paper states, "Two types of resistors appear in the GasLib instances. One results in a constant pressure loss (in the direction of the flow), which is specified by the element `pressureLoss`. The other type induces a pressure loss that continuously depends on the flow rate through the resistor as specified by the tags dragFactor and diameter."

Maybe we could refer to them as "constant" and "variable" resistors? Then I could just rename "prv" with "constant_resistor" everywhere. Or maybe "loss_resistor" or "drop_resistor?" I'm open to anything, really.

Comment (rb004f) [reply src/core/constraint.jl:43]: given that the name resistor has been agreed to with ANL, lets keep that the way it is.  Lets rename prv to loss_resistor

Comment (tasseff) [reply src/core/constraint.jl:43]: I'm fine with the name change, but thinking about this more, there's another issue: everything is in formulated in terms of squared pressure variables. However, loss resistors give a constant value for a pressure difference. So, what I currently have for a constraint right now, `loss^2 == p_i^2 - p_j^2`, is incorrect. Really, we want something that represents `loss^2 = (p_i - p_j)^2 = p_i^2 + p_j^2 - 2 * p_i * p_j`. How should we handle the bilinear term?

Comment (rb004f) [reply src/core/constraint.jl:43]: Oh, very interesting.  So the NLP and MINLP should use the constraint directly since they allow nonlinear constraints. You can either keep everything in the square space (so have 2 * \sqr(p_i) * \sqrt(p_j)), or introduced a pressure variable and have a constraint like p^2 == psqr.  I am not sure which will be better behaved.  For the MISOCP, LP, and MIP, we'll need to relax this constraint.  You could relax it by dropping it completely (which is what we are currently doing for the compressor energy constraint-- though @kaarthiksundar and I are planning to revisit this later this year) and we'll come back to it.  Indeed, since we don't yet have p variables in the LP or MIP formulation, it should just be a blank constraint for those.  

The easiest relaxation for the MISOCP might be to have psqr >= p^2 and then do McCormick on p_i*p_j. (InfrastructureModels already as these relaxations implemented).  For now, since we don't want/need p variables in the non square space, I'd say we just create them locally in the constraint definition as auxiliary variables.... and if we start to run into more and more constraints where we need p in both the linear and square space, we'll come up with an architecture which has the smarts to know when to create the linear pressure variables and when not too...

Comment (rb004f) [reply src/core/constraint.jl:43]: Another thing to think through.  Is directionality of flow a consideration here, i.e. does flow need to go in the same direction as the pressure drop? If the resistors are already one-way flow only, then we don't have to do much.  But if they are allowed to have two way flow, we'll need to enforce that.  Easy to do with the MINLP and NLP since we have direction variables.  With NLP, it is just f*(p_i-p_j) >= 0.  LP and MIP don't have p variables, so we are good.

Comment (tasseff) [reply src/core/constraint.jl]: Want to take a look at the current formulations? The constraints are `constraint_loss_resistor_pressure`. A couple things: (i) in the NLP version, I use `JuMP.@constraint(gm.model, f * (p_i - p_j) == f * pd)`, which could behave unexpectedly when `f` is zero. Not sure how to get around this while also needing to direct the pressure difference; (ii) constraints on the pressure variables at `i` and `j` are both done within a constraint for an edge and thus may be redundantly added if two loss resistors are added in series (e.g., `JuMP.@constraint(gm.model, p_i^2 == p_i_sqr)` and `JuMP.@constraint(gm.model, p_j^2 == p_j_sqr)` are added in the same constraint for one loss resistor). I can break these pressure-squaring constraints out into new constraints per junction, if you'd like.

File: test/data.jl (hunk 4 above)
Comment (tasseff) [reply test/data.jl:5]: Okay, I've added more components to the GasLib tests, which should cover all except the (empty) MIP versions of these constraints.

File: src/core/constraint.jl (hunk 1 above)
Comment (rb004f) [reply src/core/constraint.jl]: I am not entirely sure what you are trying to model with ``f * (p_i - p_j) == f * pd)''  Are you trying to make sure that direction of flow is the same as the pressure drop?  In which case, you could do ``f*(pi-pj) >= 0'' and ''pd == abs(pi - pj)''.  This will make sure the sign of f is the same as the sign of pi-pj and make sure that pressure difference matches the sign of pi - pj.  If flow can only go one way, all you need is ``pd == pi -pj'' since somewhere else you will have ``f >=0''

File: test/data.jl (hunk 4 above)
Comment (rb004f) [reply test/data.jl:5]: Ok, this looks good

File: src/core/constraint.jl (hunk 1 above)
Comment (tasseff) [reply src/core/constraint.jl]: Ah, I didn't know `abs` could be used in the constraint like that. Yes, I was trying to model pressure drop and flow directions were the same. (The components appear to be bidirectional.) This has been updated to use your suggestions.

Comment (rb004f) [reply src/core/constraint.jl]: Looks good.  I'll approve once all the checks pass.

Comment (tasseff) [reply src/core/constraint.jl]: Actually, sorry, `abs(p_i - p_j)` is not valid. However, `(p_i - p_j)^2 == pd^2` is, so we will go with that for the NLP formulation.

Comment (rb004f) [reply src/core/constraint.jl]: Really? I thought p_i, p_j, and pd were all in the space of linear pressure?  We use abs in the Weymouth formulation....

Comment (tasseff) [reply src/core/constraint.jl]: Invalid meaning you can't squeeze in a difference of two variables in an `abs` operator.

Comment (rb004f) [reply src/core/constraint.jl]: Ah, AutoDiff must not be able to handle it.  We could introduce an auxiliary variable, but your way also works,
